Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/ApromoreZKListener.java

```diff
@@ -0,0 +1,200 @@
+package org.apromore.portal;
+
+/*-
+ * #%L
+ * This file is part of "Apromore Core".
+ * %%
+ * Copyright (C) 2018 - 2021 Apromore Pty Ltd.
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.core.MediaType;
+
+import com.nimbusds.jwt.JWTParser;
+import io.netty.handler.codec.base64.Base64Decoder;
+import org.apromore.plugin.portal.PortalLoggerFactory;
+import org.apromore.portal.security.helper.JwtHelper;
+import org.apromore.portal.util.SecurityUtils;
+import org.slf4j.Logger;
+import org.zkoss.spring.SpringUtil;
+import org.zkoss.zk.ui.Execution;
+import org.zkoss.zk.ui.Session;
+import org.zkoss.zk.ui.event.Event;
+import org.zkoss.zk.ui.event.EventQueues;
+import org.zkoss.zk.ui.util.ExecutionInit;
+
+import java.security.PublicKey;
+import java.util.Base64;
+import java.util.Date;
+
+/**
+ * Callbacks during the lifecycle of ZK scopes.
+ */
+public class ApromoreZKListener implements ExecutionInit {
+
+    private static final Logger LOGGER = PortalLoggerFactory.getLogger(ApromoreZKListener.class);
+    private static final String JWT_COOKIE_NAME = "App_Auth";
+    private static final String KEY_ALIAS = "apseckey";
+
+    /**
+     * {@inheritDoc}
+     *
+     * This implementation logs the user out if the security token has expired, or refreshes it
+     * if it's still valid but close to expiry.
+     */
+    @Override
+    public void init(final Execution exec, final Execution parent) {
+        LOGGER.debug("Initialize execution {} with parent {}", exec, parent);
+
+        // If there is a parent execution, it will have already performed the required work
+        if (parent != null) {
+            return;
+        }
+
+        final HttpServletRequest httpServletRequest = (HttpServletRequest) exec.getNativeRequest();
+        final HttpServletResponse httpServletResponse = (HttpServletResponse) exec.getNativeResponse();
+
```

Comment:
The main polling for JWT token (near) expiry is put under init() method of ExecutionInit, meaning it is only executed if there is an request sent from browser to the server (https://www.zkoss.org/javadoc/8.0.0/zk/org/zkoss/zk/ui/Execution.html). If the user is idle (no requests occur) and the HTTP session times out (the session was automatically destroyed by web server and ZK), this method init() is still not executed. Now, if the user makes any action, this code here won't be executed because ZK has destroyed all desktops and UI components created before in the session (https://www.zkoss.org/wiki/ZK_Developer's_Reference/UI_Patterns/Session_Timeout_Management), the user was pointed back to the login.zul (configured in zk.xml) and asked to log in again. The state of user login on Keyloak server and any related components would be inconsistent now: they are not logged out yet because this code is not executed. Any issue?

Reply:
Does the use of ZK's Execution hook work in this case: if the user just accidentally or temporarily closes all the browser tabs of Apromore including the Portal for a while, then s/he opens the Portal again by typing in the URL box? For example, the user may open too many tabs while working (e.g. Apromore, Facebook, Youtube, news, etc.) and s/he decides to close all of them, rests for a while and then comes back to opening Apromore portal again.

Reply:
> The state of user login on Keyloak server and any related components would be inconsistent now: they are not logged out yet because this code is not executed. Any issue (e.g. is it all the cookies related to this session are removed or still there)?

As far as session consistency goes, there are two sessions that matter:
* JSESSIONID cookie, stored by reference to an HttpSession managed by Jetty (the ZK Session wraps this)
* App_Auth/Signed_App_Auth cookies, stored by value in the JWT format (includes its own expiry time)

As long as both of these sessions are current, the user can access the system.  Reasoning by cases, it doesn't seem to matter whether the cookies are cleared or not after they expire:
*  **JSESSIONID absent:**  Jetty will generate a new JSESSIONID cookie, causing the previous ZK Session to be abandoned.  The new ZK Session doesn't have a user attribute, so they're redirected to the login page.
* **JSESSIONID present:**  The identifier it contains no longer corresponds to any HttpSession Jetty has recorded (it expired after 30 minutes of inactivity).  Jetty will generate a new HttpSession and JSESSIONID cookie, and things proceed as above since the new HttpSession doesn't have a user attribute.
* **App_Auth/Signed_App_Auth absent:**  The Execution hook fails when it doesn't find the cookie and triggers sign out.
* **App_Auth/Signed_App_Auth present:**  The JWT fails validation when its expiry time is checked, triggering sign out as before.

If the Keycloak server is managing some kind of SSO session so that you only have to enter passwords or use 2FA once, that's really its business.  My understanding is that we're not currently doing anything so fancy.  Once we've got a JWT from it, we don't care whether it remembers that our browser still exists.

Reply:
> Does the use of ZK's Execution hook work in this case: if the user just accidentally or temporarily closes all the browser tabs of Apromore including the Portal for a while, then s/he opens the Portal again by typing in the URL box?

As long as the JSESSIONID and App_Auth/Signed_App_Auth cookies are still on their browser and they haven't delayed so long that either the HttpSession or JWT have expired, they ought to still be able to open the Portal, yes.

Reply:
Your explanation above is contradictory with what you do programming. In normal case, when the user clicks "Sign Out", the event listener in UserMenuController will call to ManagerService.logoutUserAllSessions() which will call to a REST endpoint to do some logout process with Keycloak server, then calls to Spring logout process and you're trying to make these processes work consistently. In the exceptional case above, you say it's fine, the Spring logout process is done but REST endpoint logout process is not done.